Userland/Libraries/LibWeb/Layout/LineBoxFragment.cpp

```diff
@@ -45,4 +45,35 @@ bool LineBoxFragment::is_atomic_inline() const
     return layout_node().is_replaced_box() || (layout_node().display().is_inline_outside() && !layout_node().display().is_flow_inside());
 }
 
+void LineBoxFragment::append_glyph_run(RefPtr<Gfx::GlyphRun> const& glyph_run, CSSPixels run_width, CSS::Direction direction)
+{
+    if (m_current_insert_direction != direction) {
+        if (direction == CSS::Direction::Rtl)
+            m_rtl_insert_position = width().to_float();
+        m_current_insert_direction = direction;
+    }
+
+    switch (direction) {
+    case CSS::Direction::Ltr:
+        for (auto& glyph : glyph_run->glyphs()) {
+            glyph.visit([&](auto& glyph) { glyph.position.translate_by(width().to_float(), 0); });
+            m_glyph_run->append(glyph);
+        }
+        break;
+    case CSS::Direction::Rtl:
+        for (auto& glyph : m_glyph_run->glyphs()) {
+            glyph.visit([&](auto& glyph) {
+                if (glyph.position.x() >= m_rtl_insert_position)
+                    glyph.position.translate_by(run_width.to_float(), 0);
+            });
+        }
+        for (auto& glyph : glyph_run->glyphs()) {
+            glyph.visit([&](auto& glyph) { glyph.position.translate_by(m_rtl_insert_position, 0); });
```

Comment:
Please test this with mixed paragraphs, which contain LTR and RTL characters. In the text bellow the aleph (looks like an X) is first, and the mem (more or less like an O) should be on the end. In the middle - you should have perfect english. I will leave the honor to you to make proper test cases :)

Both of the next examples contain the same text, but different direction of the paragraph (emulated using quote and code in markdown).

```
אאא test מממ
```

> אאא test מממ

Reply:
I've created a test case covering different combinations of mixing LRT with RTL, and have got all of them to behave the same as Firefox and Chrome:

```html
<div dir="ltr" align="left">hello test 1, 2, 3!</div>
<div dir="rtl" align="left">hello test 1, 2, 3!</div>
<div dir="ltr" align="right">hello test 1, 2, 3!</div>
<div dir="rtl" align="right">hello test 1, 2, 3!</div>

<div dir=ltr>אא aaa bbb ccc מממ</div>
<div dir=rtl>אא aaa bbb ccc מממ</div>

<div dir=ltr>אא 1 2 3 מממ</div>
<div dir=rtl>אא 1 2 3 מממ</div>

<div dir=ltr>aa....!!!</div>
<div dir=rtl>aa....!!!</div>

<div dir=ltr>حسنًا ، hello friends مرحباً أيها ا test لأصدقاء end</div>
<div dir=rtl>حسنًا ، hello friends مرحباً أيها ا test لأصدقاء end</div>
```

There is one case I know about that doesn't work, and that's brackets. It should replace them with their mirrored counterparts when flowing the text RTL. But LibUnicode doesn't currently implement mirrored characters, so I feel that might be outside of the PR.

If you know of anything big I missed, please let me know. Just note that for this PR, I'm purely focusing on the arrangement withing a single text fragment. Anything outside of that will effect layout and I feels that's beyond the scope.